src/eko/anomalous_dimensions/__init__.py

```diff
@@ -205,3 +237,270 @@ def gamma_singlet(order, n, nf):
+    # cache the s-es
+    max_weight = max(order)
+    if max_weight >= 3:
+        # here we need only S1,S2,S3,S4
+        sx = harmonics.sx(n, max_weight=max_weight + 1)
+    else:
+        sx = harmonics.sx(n, max_weight=3)
+    if order[1] >= 1:
+        sx_ns_qed = harmonics.compute_qed_ns_cache(n, sx[0])
+    # now combine
+    gamma_ns = np.zeros((order[0] + 1, order[1] + 1), np.complex_)
+    if order[0] >= 1:
+        gamma_ns[1, 0] = as1.gamma_ns(n, sx[0])
+    if order[1] >= 1:
+        gamma_ns[0, 1] = choose_ns_ad_aem1(mode, n, sx)
+    if order[0] >= 1 and order[1] >= 1:
+        gamma_ns[1, 1] = choose_ns_ad_as1aem1(mode, n, sx, sx_ns_qed)
+    # NLO and beyond
+    if order[0] >= 2:
+        if mode in [10102, 10103]:
+            gamma_ns[2, 0] = as2.gamma_nsp(n, nf, sx)
+        # To fill the full valence vector in NNLO we need to add gamma_ns^1 explicitly here
+        elif mode in [10202, 10203]:
+            gamma_ns[2, 0] = as2.gamma_nsm(n, nf, sx)
+        else:
+            raise NotImplementedError("Non-singlet sector is not implemented")
+    if order[1] >= 2:
+        gamma_ns[0, 2] = choose_ns_ad_aem2(mode, n, nf, sx, sx_ns_qed)
+    # NNLO and beyond
+    if order[0] >= 3:
+        if mode in [10102, 10103]:
+            gamma_ns[3, 0] = as3.gamma_nsp(n, nf, sx)
+        elif mode in [10202, 10203]:
+            gamma_ns[3, 0] = as3.gamma_nsm(n, nf, sx)
+    return gamma_ns
+
+
+@nb.njit(cache=True)
+def choose_ns_ad_aem1(mode, n, sx):
+    r"""
+    Select the non-singlet anomalous dimension at O(aem1) with the correct charge factor.
+
+    Parameters
+    ----------
+        mode : 10102 | 10202 | 10103 | 10203
+            sector identifier
+        n : complex
+            Mellin variable
+        nf : int
+            Number of active flavors
+
+    Returns
+    -------
+        gamma_ns : numpy.ndarray
+            non-singlet anomalous dimensions
+    """
+    if mode in [10102, 10202]:
+        return constants.eu2 * aem1.gamma_ns(n, sx)
+    if mode in [10103, 10203]:
+        return constants.ed2 * aem1.gamma_ns(n, sx)
+    else:
+        raise NotImplementedError("Non-singlet sector is not implemented")
+
+
+@nb.njit(cache=True)
+def choose_ns_ad_as1aem1(mode, n, sx, sx_ns_qed):
+    r"""
+    Select the non-singlet anomalous dimension at O(as1aem1) with the correct charge factor.
+
+    Parameters
+    ----------
+        mode : 10102 | 10202 | 10103 | 10203
+            sector identifier
+        n : complex
+            Mellin variable
+        nf : int
+            Number of active flavors
+
+    Returns
+    -------
+        gamma_ns : numpy.ndarray
+            non-singlet anomalous dimensions
+    """
+    if mode == 10102:
+        return constants.eu2 * as1aem1.gamma_nsp(n, sx, sx_ns_qed)
+    elif mode == 10103:
+        return constants.ed2 * as1aem1.gamma_nsp(n, sx, sx_ns_qed)
+    elif mode == 10202:
+        return constants.eu2 * as1aem1.gamma_nsm(n, sx, sx_ns_qed)
+    elif mode == 10203:
```

Comment:
why did you mark this resolved?

Reply:
Because the error is raised only for the order O(aem1), since it is useless to check it for all orders. If it doesn't fail at the beginning it will not fail later.

Reply:
that's not what I meant! if you can't find the `mode` either always raise an error (maybe my preferred way and as you do in the function right above this one) or return `None` (as you do here). If you decide to raise at a different level, that is just fine, but please do it consistently all over the place